### PR TITLE
Move cpu copy to aten

### PR DESCRIPTION
**Pull request by li-roy**
Stack:
&nbsp;&nbsp;&nbsp;&nbsp;:black_circle:&nbsp; **#13347 Move cpu copy to aten**&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D12850691/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #13603 [wip] Move THTensor_(copy) to aten&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D12936031/)

Differential Revision: [D12850691](https://our.internmc.facebook.com/intern/diff/D12850691/)

### CODE REVIEW

File: aten/src/ATen/Copy.h [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+namespace at {
+
+template <typename T>
+struct inter_copy_type {
+  using type = T;
+};
+
+template <>
+struct inter_copy_type<uint8_t> {
+  using type = int64_t;
```
Comment (colesbury) [inline aten/src/ATen/Copy.h]: What's this for?

**Review by colesbury — COMMENTED** [review]
Copy is pretty important speed-wise. Do you have timing numbers before/after?

File: aten/src/ATen/native/Copy.h [hunk 2]
```diff
@@ -0,0 +1,46 @@
+#pragma once
+
+#include "ATen/ATen.h"
+
+namespace at {
+namespace native {
+
+// C and C++ have a lovely set of implicit conversion rules, where casting
```
Comment (ezyang) [inline aten/src/ATen/native/Copy.h]: Err, shouldn't you delete the old copy of this comment?

Alt, give it a label `Note [Implicit conversion between signed and unsigned]` and just refer to the label here. But copy pasting the comment is a no-no.

**Comment by ezyang** [comment]
Can you post the generated code before and after?

**Comment by ezyang** [comment]
This doesn't seem like it's actually moving anything, because the old code wasn't deleted. Is there something that makes this difficult?

**Comment by ezyang** [comment]
I'm pretty sure you're going to need to vectorize this to have it run as fast as the old code; I see calls to `THVector_(copy)` in the old TH implementation, and that's our entrypoint to vectorized code.

You also probably want to call `memcpy` when no conversion happens, although if you're planning to test for converting/non-converting at the top-level, it's probably not necessary to put it directly in the code here.

**Comment by li-roy** [comment]
Regarding vectorization: that only happens in THTensor_(copy), but none of the code I'm replacing here actually uses THTensor_(copy). I'm just replacing THTensor_(copyByte), etc. which is just a tensor apply.

I can't delete the old code yet because the cuda copy code has a dependency on it.

**Comment by li-roy** [comment]
CPUCopy.cpp
before: https://gist.github.com/li-roy/a1a1addcf17c1490315d8c2bcc9e8937
after: https://gist.github.com/li-roy/a1bc93e8c59595c24c70bfdc9d5594b9

as you can see, i'm only replacing instances of THTensor_(copyByte), etc in this patch.

**Comment by ezyang** [comment]
@li-roy, are you planning to call into ATen from THC so you can eliminate the TH copy? If so, I'll wait until you add that change to review.

**Comment by li-roy** [comment]
Okay, this is ready for review now.

**Comment by ezyang** [comment]
It doesn't look like `THTensor_(copy)` is deleted yet. Shall I proceed under the assumption that you are not deleting it?

File: aten/src/TH/THTensor.hpp [hunk 3]
```diff
@@ -9,6 +9,11 @@
 #include <atomic>
 #include <ATen/ATen.h>
 
+inline at::Tensor THTensor_wrap(THTensor* tensor) {
```
Comment (ezyang) [inline aten/src/TH/THTensor.hpp]: Needs a doc; specifically, the doc should say that `THTensor* tensor` stays valid even after this call.

File: aten/src/THC/generic/THCTensorCopy.cpp [hunk 4]
```diff
@@ -94,9 +96,11 @@ void THTensor_(copyCuda)(THCState *state, THTensor *self, struct THCTensor *src)
       THTensor *srcf = THTensor_(newWithSize)(src->sizes(), {});          \
                                                                           \
       THTensor_(copyCuda)(state, srcf, src);                              \
-      TH_CONCAT_4(TH,TYPEC,Tensor_copy,Real)(self, srcf);                 \
+      at::Tensor srcf_wrap = THTensor_wrap(srcf);                         \
+      at::Tensor self_wrap = THTensor_wrap(self);                         \
+      at::_copy_(self_wrap, srcf_wrap);                                    \
                                                                           \
-      c10::raw::intrusive_ptr::decref(srcf);                                              \
+      c10::raw::intrusive_ptr::decref(srcf);                              \
```
Comment (ezyang) [inline aten/src/THC/generic/THCTensorCopy.cpp]: nit: Since `srcf` is never used after here, you don't need to incref/decref; you can just steal it entirely, saving an incref/decref. This is unlikely to show up in a singlethreaded benchmark, but we should practice good hygiene.

Actually, going one step further, `srcf` is allocated using `THTensor_(newWithSize)`, but you could use an equivalent ATen function `at::empty(src->sizes())` to have the same effect, and get a `at::Tensor` while you're at it.

**Review by ezyang — APPROVED** [review]
Pleasingly simple!

**Comment by li-roy** [comment]
Did some simple benchmarking with this example:

```
import torch

def a():
  b = torch.randn(10,10)
  c = torch.randn(10,10,dtype=torch.double)
  b.copy_(c)
  
%timeit a()
```

Results are consistently the same before and after this PR. Some examples given below.
100 elements, float: 8.81 µs ± 99.7 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
1M elements, float: 19.3 ms ± 329 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
100 elements, double to float: 17.3 µs ± 145 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
1M elements, double to float: 82.7 ms ± 1.7 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
